From 3d4afc2273f4e8bd8c312118171bc5a95b644505 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:39:01 -0400
Subject: [ethernet] i40e: Fix a bug in ethtool for FD drop packet filter action

Message-id: <1409658042-22689-181-git-send-email-sassmann@redhat.com>
Patchwork-id: 90376
O-Subject: [RHEL7.1 PATCH 180/281] i40e: Fix a bug in ethtool for FD drop packet filter action
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Anjali Singhai Jain <anjali.singhai@intel.com>

A drop action comes down as a ring_cookie value, so allow it as
a special value that can be used to configure destination control.

Also fix the output to filter read command accordingly.

Change-ID: I9956723cee42f3194885403317dd21ed4a151144
Signed-off-by: Anjali Singhai Jain <anjali.singhai@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 387ce1a97d703ae62032f802313113b856de303b)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_ethtool.c b/drivers/net/ethernet/intel/i40e/i40e_ethtool.c
index 0c935c5..b08290d 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_ethtool.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_ethtool.c
@@ -1259,7 +1259,11 @@ static int i40e_get_ethtool_fdir_entry(struct i40e_pf *pf,
  fsp->h_u.tcp_ip4_spec.pdst = rule->src_port;
  fsp->h_u.tcp_ip4_spec.ip4src = rule->dst_ip[0];
  fsp->h_u.tcp_ip4_spec.ip4dst = rule->src_ip[0];
- fsp->ring_cookie = rule->q_index;
+
+ if (rule->dest_ctl == I40E_FILTER_PROGRAM_DESC_DEST_DROP_PACKET)
+  fsp->ring_cookie = RX_CLS_FLOW_DISC;
+ else
+  fsp->ring_cookie = rule->q_index;
 
  return 0;
 }
@@ -1563,7 +1567,8 @@ static int i40e_add_fdir_ethtool(struct i40e_vsi *vsi,
   return -EINVAL;
  }
 
- if (fsp->ring_cookie >= vsi->num_queue_pairs)
+ if ((fsp->ring_cookie != RX_CLS_FLOW_DISC) &&
+     (fsp->ring_cookie >= vsi->num_queue_pairs))
   return -EINVAL;
 
  input = kzalloc(sizeof(*input), GFP_KERNEL);
-- 
1.7.1